README.md

## About

This repository contains four Python scripts that can be used to
generate a movie database for the trivia game [MovieWar][].

You probably don't need to use them, the generated JSON files are
already available in the [Releases][] tab.

[MovieWar]: https://github.com/Beluki/MovieWar
[Releases]: https://github.com/Beluki/MovieWarDBGen/releases

## How it works

The first script, [00 download freebase.py][], uses the [Freebase API][]
to download information for about 9200 movies, using Brad Bourland's
[top movies list][] as a source.

[Freebase API]: https://developers.google.com/freebase/
[top movies list]: http://www.nytimes.com/2010/04/18/movies/18bourland.html?_r=0

This is a great starting point for a trivia game. It's a big enough
number of movies to avoid repetition. All the movies are reasonably
popular. No Bollywood or obscure titles.

Sample output:

```text
Total requests: 1 Total movies: 3000
Total requests: 2 Total movies: 6000
Total requests: 3 Total movies: 9000
Total requests: 4 Total movies: 9174
All data gathered, saving...
```

Freebase is a great resource, but nothing is perfect.

The second script, [01 convert freebase.py][], sanitizes the list.
It removes movies with no title or release date and sorts the movies by name.
It also checks the release date to be between 1900 and 2000, which should be
true for all the movies in the list.

5 of the 9174 movies are removed due to invalid name/dates.

Sample output:

```text
Invalid year range: 2002 for movie: Stevie, skipping...
Empty date, skipping movie: Siddhartha
Invalid year range: 2003 for movie: Johnny Thunders: What About Me, skipping...
Invalid year range: 2004 for movie: The Big Bounce, skipping...
Invalid year range: 2004 for movie: Grizzly Falls, skipping...
```

The third script, [02 match omdb.py][], uses the [OMDB API][] to check
that the movie names/dates are actually correct. Only the movies that have the
exact same title and date on both Freebase and OMDB are kept.

[OMDB API]: http://www.omdbapi.com

In a trivia game, quality is more important than quantity.
Using multiple sources of information guarantees correctness.
From the 9169 movies, 7601 match in both databases.

Sample output:

```text
1 - ok: 0 miss: 1 - movie title mismatch.
2 - ok: 0 miss: 2 - movie title mismatch.
3 - ok: 1 miss: 2 - movie ok.
4 - ok: 2 miss: 2 - movie ok.
5 - ok: 3 miss: 2 - movie ok.
6 - ok: 4 miss: 2 - movie ok.
7 - ok: 4 miss: 3 - movie title mismatch.
8 - ok: 4 miss: 4 - movie title mismatch.
9 - ok: 5 miss: 4 - movie ok.
10 - ok: 6 miss: 4 - movie ok.
```

The fourth script, [03 collapse years.py][], looks for duplicate movie names
and years. For example, it converts this:

```json
{"name": "Jane Eyre", "year": "2011"}
{"name": "Jane Eyre", "year": "1996"}
```

Into this:

```json
{"name": "Jane Eyre", "years": ["2011", "1996"]}
```

It also checks for duplicate years.
Only one movie in the entire dataset produces this error.

Sample output:

```text
Skipping duplicate year for movie: A Doll's House...
```

[00 download freebase.py]: Source/00%20download%20freebase.py
[01 convert freebase.py]: Source/01%20convert%20freebase.py
[02 match omdb.py]: Source/02%20match%20omdb.py
[03 collapse years.py]: Source/03%20collapse%20years.py

## Portability

Information and error messages are written to stdout and stderr
respectively, using the current platform newline format and encoding.

Note that since the scripts were a one-off effort (once we have the JSON
there's no need to run them again), they actually do very little to no
error checking.

There are no options or arguments.

The output JSON is written as UTF-8 without BOM, using Unix newlines.

I wrote and ran them on Windows 7 x86-64, using Python 3.5.0 and
requests 2.9.1.

## Status

This program is finished!

Those scripts served their purpose, generating a good, basic database
for MovieWar. I plan no further development on them.

## License

Like all my hobby projects, this is Free Software. See the [Documentation][]
folder for more information. No warranty though.

[Documentation]: Documentation



Documentation/Changelog

CHANGELOG

* 2016/02/03:

    - Revised. Working on Python 3.5.0, requests 2.9.1.



Source/02 match omdb.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Checks each movie date from "01 convert freebase.json"
against the OMDB API, removing movies where the title and the date
do not match.

Running this script can take one, two hours or more, since there
are about 9200 movies to check.

After running, two output files are created:
   02 match omdb.json - the movies that matched.
   02 mismatch omdb.json - the movies that didn't match, for logging.
"""


import json
import sys

from urllib.parse import quote


# Information and error messages:

def outln(line):
    """ Write 'line' to stdout, using the platform encoding and newline format. """
    print(line, flush = True)


def errln(line):
    """ Write 'line' to stderr, using the platform encoding and newline format. """
    print('02 match omdb.py: error:', line, file = sys.stderr, flush = True)


# Non-builtin imports:

try:
    import requests

except ImportError:
    errln('This script requires the following modules:')
    errln('requests 2.7.0+ - <https://pypi.python.org/pypi/requests>')
    sys.exit(1)


# Making requests:

OMDB_API = 'http://www.omdbapi.com/'


def request_json(title, year):
    """
    Make a OMDB request for a given movie title, released on a particular year.
    """
    query = OMDB_API + '?t=' + quote(title) + '&?y=' + year
    request = requests.get(query)

    return request.json()


# Writing:

def write_results(results, filename):
    """
    Write the results from our checks to a given file.
    Each record is written in a separate line.
    """
    with open(filename, mode = 'wb') as descriptor:
        for result in results:
            jsonbytes = json.dumps(result).encode('utf-8')
            descriptor.write(jsonbytes)
            descriptor.write(b'\n')


# Entry point:

def main():
    input_movies = open('01 convert freebase.json', 'r', encoding = 'utf-8').read().splitlines()

    match_movies = []
    mismatch_movies = []

    total_match = 0
    total_mismatch = 0

    for index, line in enumerate(input_movies, 1):
        movie = json.loads(line)

        title = movie['name']
        year = movie['year']

        omdbjson = request_json(title, year)

        # validate title match:
        omdbtitle = omdbjson.get('Title', '') or ''
        omdbtitle = omdbtitle.strip()

        if omdbtitle == '' or omdbtitle != title:
            mismatch_movies.append(movie)
            total_mismatch += 1
            outln('{} - ok: {} miss: {} - movie title mismatch.'.format(index, total_match, total_mismatch))
            continue

        # validate year match:
        omdbyear = omdbjson.get('Year', '') or ''
        omdbyear = omdbyear.strip()
        omdbyear = omdbyear[:4]

        if omdbyear == '' or omdbyear != year:
            mismatch_movies.append(movie)
            total_mismatch += 1
            outln('{} - ok: {} miss: {} - movie year mismatch.'.format(index, total_match, total_mismatch))
            continue

        # ok in both:
        match_movies.append(movie)
        total_match += 1
        outln('{} - ok: {} miss: {} - movie ok.'.format(index, total_match, total_mismatch))

    write_results(match_movies, '02 match omdb.json')
    write_results(mismatch_movies, '02 mismatch omdb.json')


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass



Source/03 collapse years.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Checks movies from "02 match omdb.json" and collapses movies with the
same title but different years into one record with the years
as a list.

e.g.

Input:

{"name": "Jane Eyre", "year": "2011"}
{"name": "Jane Eyre", "year": "1996"}

Output:

{"name": "Jane Eyre", "years": ["2011", "1996"]}

The output is saved as: "03 collapse years.json".
It also skips duplicate years.
"""


import json
import sys

from collections import OrderedDict


# Information and error messages:

def outln(line):
    """ Write 'line' to stdout, using the platform encoding and newline format. """
    print(line, flush = True)


def errln(line):
    """ Write 'line' to stderr, using the platform encoding and newline format. """
    print('03 collapse years.py: error:', line, file = sys.stderr, flush = True)


# Entry point:

def main():
    input_movies = open('02 match omdb.json', 'r', encoding = 'utf-8').read().splitlines()
    output_movies = OrderedDict()

    for line in input_movies:
        movie = json.loads(line)

        name = movie['name']
        year = movie['year']

        output_movies.setdefault(name, [])

        if year in output_movies[name]:
            outln('Skipping duplicate year for movie: {}...'.format(name))
        else:
            output_movies[name].append(year)

    # save, years first:
    with open('03 collapse years.json', mode = 'wb') as descriptor:
        for name, years in output_movies.items():
            movie = OrderedDict()

            movie['years'] = years
            movie['name'] = name

            jsonbytes = json.dumps(movie).encode('utf-8')
            descriptor.write(jsonbytes)
            descriptor.write(b'\n')


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass



Source/01 convert freebase.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Filters the output from "00 download freebase.json" to match
the MovieWar JSON format. Removes movies with no name
or release date, etc...
"""


import json
import sys

from operator import itemgetter


# Information and error messages:

def outln(line):
    """ Write 'line' to stdout, using the platform encoding and newline format. """
    print(line, flush = True)


def errln(line):
    """ Write 'line' to stderr, using the platform encoding and newline format. """
    print('01 convert freebase.py: error:', line, file = sys.stderr, flush = True)


# Entry point:

def main():
    input_movies = open('00 download freebase.json', 'r', encoding = 'utf-8').read().splitlines()
    output_movies = []

    for index, line in enumerate(input_movies):
        movie = json.loads(line)

        # validate name:
        name = movie.get('name') or ''
        name = name.strip()

        if name == '':
            outln('Empty name, skipping movie: {}'.format(index))
            continue

        # validate date:
        date = movie.get('initial_release_date') or ''
        date = date.strip()

        if date == '':
            outln('Empty date, skipping movie: {}'.format(name))
            continue

        # 'initial_release_date' is in the following format: (YYYY-MM-DD)
        # for the trivia we need the year (and sometimes only the year is present)
        year = date[:4]

        # validate the year:
        try:
            year_as_number = int(year)

            # all the movies in the 9200 movie set are supposed to be
            # between 1900 and 2000:
            if not (1900 <= year_as_number <= 2000):
                outln('Invalid year range: {} for movie: {}, skipping...'.format(year, name))
                continue

        except ValueError:
            outln('Invalid date: {}, for movie: {}'.format(date, name))
            continue

        # create a new JSON value, using the trivia format:
        movie = { 'name': name, 'year': year }
        output_movies.append(movie)

    # sort the movies by name:
    output_movies = sorted(output_movies, key = itemgetter('name'))

    # save:
    with open('01 convert freebase.json', 'wb') as descriptor:
        for movie in output_movies:
            jsonbytes = json.dumps(movie).encode('utf-8')
            descriptor.write(jsonbytes)
            descriptor.write(b'\n')


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass



Source/00 download freebase.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Downloads the base movie cache dataset from Freebase
using Brad Bourland 9200 top movies list as a source.

Running this script usually takes about 1 to 10 minutes.
"""


import json
import time
import sys


# Information and error messages:

def outln(line):
    """ Write 'line' to stdout, using the platform encoding and newline format. """
    print(line, flush = True)


def errln(line):
    """ Write 'line' to stderr, using the platform encoding and newline format. """
    print('00 download freebase.py: error:', line, file = sys.stderr, flush = True)


# Non-builtin imports:

try:
    import requests

except ImportError:
    errln('This script requires the following modules:')
    errln('requests 2.7.0+ - <https://pypi.python.org/pypi/requests>')
    sys.exit(1)


# MQL READ API url:

MQLREAD_URL = 'https://www.googleapis.com/freebase/v1/mqlread?query='


# QUERY:

MQLREAD_QUERY = '''
[{
    "type": "/film/film",
    "b:type": "/award/ranked_item",

    "/award/ranked_item/appears_in_ranked_lists": [{
        "list": "The Movie List: the first 9200"
    }],

    "name": null,
    "initial_release_date": null,
    "directed_by": [],

    "limit": 3000
}]&html_escape=false&cursor'''.replace('\n', '')


# Requesting and manipulating data:

def request_json(cursor = None):
    """
    Make a Freebase request asking for more movies.
    """
    query = MQLREAD_QUERY

    if cursor is not None:
        query = query + '=' + cursor

    request = requests.get(MQLREAD_URL + query)
    return request.json()


def write_results(results, filename):
    """
    Write the results from a request to a given file.
    Each record is written in a separate line.
    """
    with open(filename, mode = 'wb') as descriptor:
        for result in results:
            jsonbytes = json.dumps(result).encode('utf-8')
            descriptor.write(jsonbytes)
            descriptor.write(b'\n')


# Entry point:

def main():
    results = []
    cursor = None
    total_requests = 0

    while True:

        # request more movies, on an error wait 5 minutes
        # and try again, to respect API usage limits:
        while True:
            total_requests += 1
            data = request_json(cursor)

            if not 'result' in data:
                errorjson = json.dumps(data, indent = 4, sort_keys = True)

                errln('Error while requesting data! Response follows:')
                errln(errorjson)

                errln('Retrying in 5 minutes...')
                time.sleep(300)

            # got the data, stop:
            else:
                break

        results += data['result']
        cursor = data['cursor']

        outln('Total requests: {} Total movies: {}'.format(total_requests, len(results)))

        # wait 10 seconds to make the next request:
        time.sleep(10)

        if not cursor:
            outln('All data gathered, saving...')
            write_results(results, '00 download freebase.json')
            break


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass

